{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a43097c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pathlib\n",
    "import pickle\n",
    "import itertools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17a5ee8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import autoreload\n",
    "import evotsc\n",
    "import evotsc_plot\n",
    "autoreload.reload(evotsc)\n",
    "autoreload.reload(evotsc_plot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7cca8c7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Constants\n",
    "# Population\n",
    "nb_indivs = 100\n",
    "nb_genes = 60\n",
    "\n",
    "# Genome\n",
    "intergene = 2000\n",
    "gene_length = 1000\n",
    "interaction_dist = 5000\n",
    "interaction_coef = 0.3\n",
    "sigma_basal = -0.06\n",
    "sigma_opt = -0.06\n",
    "epsilon = 0.03\n",
    "default_basal_expression = 0.5\n",
    "nb_eval_steps = 51\n",
    "\n",
    "# Environment\n",
    "sigma_A = 0.1\n",
    "sigma_B = -0.1\n",
    "\n",
    "# Mutations\n",
    "inversion_poisson_lam = 2.0\n",
    "intergene_poisson_lam = 0.0 #2.0\n",
    "intergene_mutation_var = 0.0 #1e1\n",
    "basal_sc_mutation_prob = 0.0 #1e-1\n",
    "basal_sc_mutation_var = 0.0 #1e-4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7b2f4e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_types = ['AB', 'A', 'B'] # Name of each gene type\n",
    "gene_type_color = ['tab:blue', 'tab:red', 'tab:green'] #AB, A, B\n",
    "dpi = 300"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4dce3be3",
   "metadata": {},
   "outputs": [],
   "source": [
    "exp_path = pathlib.Path('/Users/theotime/Desktop/evotsc/inter-5k/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0abf279c",
   "metadata": {},
   "outputs": [],
   "source": [
    "rng = np.random.default_rng(seed=123456)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3adca56",
   "metadata": {},
   "source": [
    "# Helper functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d4ca612",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_params(exp_path):\n",
    "    rep_dirs = sorted([d for d in exp_path.iterdir() if (d.is_dir() and d.name.startswith(\"rep\"))])\n",
    "    \n",
    "    with open(rep_dirs[0].joinpath('params.txt'), 'r') as params_file:\n",
    "        param_lines = params_file.readlines()\n",
    "        \n",
    "    params = {}\n",
    "    for line in param_lines:\n",
    "        param_name = line.split(':')[0]\n",
    "        if param_name == 'commit':\n",
    "            param_val = line.split(':')[1].strip()\n",
    "        else:\n",
    "            param_val = float(line.split(':')[1])\n",
    "        \n",
    "        params[param_name] = param_val\n",
    "        \n",
    "    return params"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "759acb20",
   "metadata": {},
   "outputs": [],
   "source": [
    "params = get_params(exp_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df01c99a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_best_indiv(rep_path, gen):\n",
    "    \n",
    "    with open(rep_path.joinpath(f'pop_gen_{gen:06}.evotsc'), 'rb') as save_file:\n",
    "        pop_rep = pickle.load(save_file)\n",
    "        \n",
    "    pop_rep.evaluate()\n",
    "    \n",
    "    best_fit = 0\n",
    "    best_indiv = pop_rep.individuals[0]\n",
    "    \n",
    "    try:\n",
    "        for indiv in pop_rep.individuals:\n",
    "            if indiv.fitness > best_fit:\n",
    "                best_fit = indiv.fitness\n",
    "                best_indiv = indiv\n",
    "    except AttributeError: # In the neutral control, individuals are not evaluated so there is no fitness field\n",
    "        pass\n",
    "    \n",
    "    return best_indiv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a1ef1dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_expr(indiv, sigma_env, plot_title=None, plot_name=None):\n",
    "    \n",
    "    if not indiv.already_evaluated:\n",
    "        indiv.evaluate(sigma_env, sigma_env)\n",
    "    \n",
    "    # Plot only environment A\n",
    "    temporal_expr = indiv.run_system(sigma_env)\n",
    "\n",
    "    nb_genes, nb_steps = temporal_expr.shape\n",
    "\n",
    "    colors = mpl.cm.get_cmap('viridis', nb_genes)(range(nb_genes))\n",
    "\n",
    "    plt.figure(figsize=(6, 4), dpi=dpi)\n",
    "\n",
    "    plt.ylim(-0.05, 1.05)\n",
    "\n",
    "    for gene in range(nb_genes):\n",
    "        linestyle = 'solid' if indiv.genes[gene].orientation == 0 else 'dashed'\n",
    "        plt.plot(temporal_expr[indiv.genes[gene].id, :],\n",
    "                 linestyle=linestyle,\n",
    "                 linewidth=2,\n",
    "                 color=colors[indiv.genes[gene].id],\n",
    "                 label=f'Gene {indiv.genes[gene].id}')\n",
    "\n",
    "    plt.grid(linestyle=':')\n",
    "    plt.xlabel('Time', fontsize='large')\n",
    "    plt.ylabel('Expression level', fontsize='large')\n",
    "\n",
    "    if plot_title:\n",
    "        plt.title(plot_title)\n",
    "\n",
    "    plt.tight_layout()\n",
    "    if plot_name:\n",
    "        plt.savefig(plot_name + '.pdf', dpi=300, bbox_inches='tight')\n",
    "    plt.show()\n",
    "    plt.close()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8117820",
   "metadata": {},
   "source": [
    "# Plot the genome of an individual, with the final supercoiling level at each position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7669456f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute the final supercoiling level at positions `positions` under external supercoiling `sigma`\n",
    "def compute_final_sc_at(indiv, sigma, positions) -> np.ndarray:\n",
    "    gene_positions, genome_size = indiv.compute_gene_positions()\n",
    "\n",
    "    nb_pos = len(positions)\n",
    "    sc_tsc = np.zeros(nb_pos)\n",
    "\n",
    "    # Run the individual\n",
    "    if not indiv.already_evaluated:\n",
    "        indiv.evaluate(sigma, sigma)\n",
    "        \n",
    "    temporal_expr = indiv.run_system(sigma)\n",
    "    gene_expr = temporal_expr[:, -1]\n",
    "    \n",
    "    for i_pos, x in enumerate(positions):\n",
    "        \n",
    "        pos_tsc = 0.0\n",
    "        \n",
    "        for i_gene, gene in enumerate(indiv.genes):\n",
    "\n",
    "            # We compute the influence of gene i at position x\n",
    "\n",
    "            pos_1_minus_x = gene_positions[i_gene] - x\n",
    "            pos_x_minus_1 = - pos_1_minus_x\n",
    "\n",
    "            ## On veut savoir si le gène 1 est avant la position x ou après\n",
    "            # Avant : -------1--x-------- ou -x---------------1-\n",
    "            # Après : -------x--1-------- ou -1---------------x-\n",
    "\n",
    "            if pos_1_minus_x < 0: # -------1--2-------- ou -1---------------2-\n",
    "                if pos_x_minus_1 < genome_size + pos_1_minus_x: # -------1--2--------\n",
    "                    distance = pos_x_minus_1\n",
    "                    i_before_x = True\n",
    "                else: # -1---------------2-\n",
    "                    distance = genome_size + pos_1_minus_x\n",
    "                    i_before_x = False\n",
    "\n",
    "            else: # -------2--1-------- ou -2---------------1-\n",
    "                if pos_1_minus_x < genome_size + pos_x_minus_1: # -------2--1--------\n",
    "                    distance = pos_1_minus_x\n",
    "                    i_before_x = False\n",
    "                else:\n",
    "                    distance = genome_size + pos_x_minus_1\n",
    "                    i_before_x = True\n",
    "\n",
    "            # Exit early if genes are too far\n",
    "            if distance > indiv.interaction_dist:\n",
    "                continue\n",
    "\n",
    "            if i_before_x:\n",
    "                if gene.orientation == 1: # i lagging : +\n",
    "                    sign_1_on_x = +1\n",
    "                else:\n",
    "                    sign_1_on_x = -1\n",
    "            else:\n",
    "                if gene.orientation == 0: # i leading : +\n",
    "                    sign_1_on_x = +1\n",
    "                else:\n",
    "                    sign_1_on_x = -1\n",
    "\n",
    "            # Here, we know that distance <= self.interaction_dist\n",
    "            strength = 1.0 - distance/indiv.interaction_dist\n",
    "\n",
    "            # Supercoiling variations are additive\n",
    "            pos_tsc += sign_1_on_x * strength * gene_expr[i_gene]\n",
    "            \n",
    "        sc_tsc[i_pos] = pos_tsc\n",
    "\n",
    "\n",
    "    return sc_tsc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5c7d4f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_genome_and_tsc(indiv, sigma, show_types=True, print_ids=False, name=None):\n",
    "\n",
    "    # Compute gene positions\n",
    "    gene_pos, genome_length = indiv.compute_gene_positions()\n",
    "\n",
    "    # Plot\n",
    "    pos_rect = [0, 0, 1, 1]\n",
    "    fig = plt.figure(figsize=(9,9), dpi=dpi)\n",
    "    ax = fig.add_axes(pos_rect)\n",
    "\n",
    "    rect_width = 0.04\n",
    "    rect_height = 0.1\n",
    "\n",
    "    ax.set_xlim(-1.2, 1.2)\n",
    "    ax.set_ylim(-1.2, 1.2)\n",
    "    circle = plt.Circle(xy=(0, 0), radius=1, linestyle='-', fill=False)\n",
    "    ax.add_patch(circle)\n",
    "    ax.set_axis_off()\n",
    "\n",
    "    seq_colors = colors = mpl.cm.get_cmap('viridis', indiv.nb_genes)(range(indiv.nb_genes))\n",
    "    \n",
    "    ## Plot the genes themselves\n",
    "    for i_gene, gene in enumerate(indiv.genes):\n",
    "        pos_angle = 360 * gene_pos[i_gene] / genome_length\n",
    "        orient_angle = 360 - pos_angle\n",
    "        pos_rad = np.radians(pos_angle)\n",
    "        orient_rad = np.radians(orient_angle)\n",
    "\n",
    "        ## Plot the gene rectangle\n",
    "\n",
    "        x0 = (1.0 - rect_height / 2.0) * np.sin(pos_rad)\n",
    "        y0 = (1.0 - rect_height / 2.0) * np.cos(pos_rad)\n",
    "\n",
    "\n",
    "        if gene.orientation == 0:\n",
    "            final_width = rect_width\n",
    "        else:\n",
    "            final_width = -rect_width\n",
    "\n",
    "        if show_types:\n",
    "            gene_color = gene_type_color[gene.gene_type]\n",
    "        else:\n",
    "            gene_color = seq_colors[i_gene]\n",
    "\n",
    "        rect = plt.Rectangle(xy=(x0, y0),\n",
    "                             width=final_width,\n",
    "                             height=rect_height,\n",
    "                             angle=orient_angle, #in degrees anti-clockwise about xy.\n",
    "                             facecolor=gene_color,\n",
    "                             edgecolor='black',\n",
    "                             label=f'Gene {i_gene}')\n",
    "\n",
    "        ax.add_patch(rect)\n",
    "\n",
    "        ## Plot the orientation bar and arrow\n",
    "\n",
    "        # Bar\n",
    "        x_lin = (1.0 + (np.array([0.5, 1.0])) * rect_height) * np.sin(pos_rad)\n",
    "        y_lin = (1.0 + (np.array([0.5, 1.0])) * rect_height) * np.cos(pos_rad)\n",
    "\n",
    "        ax.plot(x_lin, y_lin, color='black', linewidth=1)\n",
    "\n",
    "        # Arrow\n",
    "        dx_arr = rect_width * np.cos(pos_rad) / 3.0\n",
    "        dy_arr = - rect_width * np.sin(pos_rad) / 3.0\n",
    "\n",
    "        if gene.orientation == 1: # Reverse\n",
    "            dx_arr, dy_arr = -dx_arr, -dy_arr\n",
    "\n",
    "        ax.arrow(x_lin[1], y_lin[1], dx_arr, dy_arr, head_width=0.02, color='black')\n",
    "\n",
    "        ## Print gene ID\n",
    "        if print_ids and (i_gene % 5 == 0):\n",
    "            ha = 'left'\n",
    "            if gene.orientation == 1:\n",
    "                ha = 'right'\n",
    "            ax.text(x=0.92*x0, y=0.92*y0, s=f'{i_gene}',\n",
    "                    rotation=orient_angle, ha=ha, va='bottom', rotation_mode='anchor',\n",
    "                    fontsize=15)\n",
    "\n",
    "    ## Plot local supercoiling along the genome, at the end of the individual's lifecycle\n",
    "    sc_ax = fig.add_axes(pos_rect, projection='polar', frameon=False)\n",
    "    sc_ax.set_ylim(0, 1)\n",
    "\n",
    "    n = 1000  # the number of data points\n",
    "\n",
    "    # theta values (see \n",
    "    # https://matplotlib.org/devdocs/gallery/images_contours_and_fields/pcolormesh_grids.html)\n",
    "    # To have the crisp version: put n+1 in theta and [data] as the 3rd argument of pcolormesh()\n",
    "    # To have the blurry version: put n in theta and [data, data] ----------------------------\n",
    "    theta = np.linspace(0, 2 * np.pi, n)\n",
    "    radius = np.linspace(.6, .72, 2)\n",
    "\n",
    "    #data = np.array([theta[:-1]]) #np.array([np.random.random(n) * 2 * np.pi])\n",
    "    positions = np.linspace(0, genome_length, n, dtype=int)\n",
    "    data = compute_final_sc_at(indiv, sigma, positions)\n",
    "    \n",
    "    norm = mpl.colors.Normalize(-2.0, 2.0) # Extremum values for the SC level \n",
    "    \n",
    "    data = -data # Reverse data to get blue = positive and red = negative SC\n",
    "    sc_ax.pcolormesh(theta, radius, [data, data], shading='gouraud',\n",
    "                     norm=norm, cmap=plt.get_cmap('seismic'))\n",
    "    sc_ax.set_yticklabels([])\n",
    "    sc_ax.set_xticklabels([])\n",
    "    #sc_ax.spines['polar'].set_visible(False)\n",
    "    sc_ax.set_theta_zero_location('N')\n",
    "    sc_ax.set_theta_direction('clockwise')\n",
    "\n",
    "    ## Plot the legend\n",
    "    if show_types:\n",
    "        patches = [mpl.patches.Patch(facecolor=color, edgecolor='black', label=label)\n",
    "                   for color, label in zip(gene_type_color, gene_types)]\n",
    "        ax.legend(handles=patches, title='Gene type', loc='center',\n",
    "                  fontsize=15, title_fontsize=15)\n",
    "\n",
    "    line_len = np.pi*indiv.interaction_dist/genome_length\n",
    "    if show_types:\n",
    "        line_y = -0.3\n",
    "    else:\n",
    "        line_y = -0.1\n",
    "    ax.plot([-line_len, line_len], [line_y, line_y],\n",
    "             color='black',\n",
    "             linewidth=1)\n",
    "    ax.text(0, line_y - 0.07, 'Gene interaction distance', ha='center', fontsize=15)\n",
    "    \n",
    "    if name:\n",
    "        plt.savefig(name, dpi=300, bbox_inches='tight')\n",
    "\n",
    "    plt.show()\n",
    "\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6a7af11",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_random_indiv(nb_genes=nb_genes, nb_mut=0):\n",
    "    genes = evotsc.Gene.generate(intergene=intergene,\n",
    "                                 length=gene_length,\n",
    "                                 nb_genes=nb_genes,\n",
    "                                 default_basal_expression=default_basal_expression,\n",
    "                                 rng=rng)\n",
    "    \n",
    "    indiv = evotsc.Individual(genes=genes,\n",
    "                              interaction_dist=interaction_dist,\n",
    "                              interaction_coef=interaction_coef,\n",
    "                              sigma_basal=sigma_basal,\n",
    "                              sigma_opt=sigma_opt,\n",
    "                              epsilon=epsilon,\n",
    "                              rng=rng)\n",
    "    \n",
    "    mutation = evotsc.Mutation(basal_sc_mutation_prob=basal_sc_mutation_prob,\n",
    "                               basal_sc_mutation_var=basal_sc_mutation_var,\n",
    "                               intergene_poisson_lam=intergene_poisson_lam,\n",
    "                               intergene_mutation_var=intergene_mutation_var,\n",
    "                               inversion_poisson_lam=inversion_poisson_lam)\n",
    "    \n",
    "    for i_mut in range(nb_mut):\n",
    "        indiv.mutate(mutation)\n",
    "    \n",
    "    return indiv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "691fcc9e",
   "metadata": {},
   "source": [
    "# Random individual: influence of environmental supercoiling on final gene expression levels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d62cc7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_extended_phenotype(indiv, sigmas):\n",
    "\n",
    "    # Initialize the individual (compute the inter matrix)\n",
    "    indiv.evaluate(0, 0)\n",
    "\n",
    "    nb_sigmas = len(sigmas)\n",
    "    ext_phen = np.zeros((indiv.nb_genes, nb_sigmas))\n",
    "\n",
    "    for i_sigma, sigma_env in enumerate(sigmas):\n",
    "        temporal_expr = indiv.run_system(sigma_env)\n",
    "        for i_gene, gene in enumerate(indiv.genes):\n",
    "            ext_phen[i_gene, i_sigma] = temporal_expr[i_gene, -1]\n",
    "    \n",
    "    return ext_phen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e653f8d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "init_indiv = make_random_indiv(nb_genes=20, nb_mut=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3bab58b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_expr(init_indiv, sigma_env=0, plot_title='Random individual', plot_name='random_gene_expr')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81308ce6",
   "metadata": {},
   "outputs": [],
   "source": [
    "evotsc_plot.plot_genome_and_tsc(init_indiv, sigma=0, name='random_genome_and_tsc.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6aafd27",
   "metadata": {},
   "outputs": [],
   "source": [
    "# See how gene activity levels depend on environmental supercoiling\n",
    "def plot_activity_sigma_all_genes(indiv, plot_title=None, plot_name=None):\n",
    "\n",
    "    colors = mpl.cm.get_cmap('viridis', indiv.nb_genes)(range(indiv.nb_genes))\n",
    "    \n",
    "    nb_sigmas = 100\n",
    "    \n",
    "    plt.figure(figsize=(6, 4), dpi=dpi)\n",
    "    plt.xlabel('Environment supercoiling')\n",
    "    plt.ylabel('Final gene activity')\n",
    "    plt.ylim(-0.05, 1.05)\n",
    "    plt.grid(linestyle=':')\n",
    "    if plot_title:\n",
    "        plt.title(plot_title)\n",
    "\n",
    "    sigmas = np.linspace(-0.25, 0.15, nb_sigmas)\n",
    "    ext_phen = compute_extended_phenotype(indiv, sigmas)\n",
    "\n",
    "    for i_gene in range(indiv.nb_genes):\n",
    "        linestyle = 'solid' if indiv.genes[i_gene].orientation == 0 else 'dashed'\n",
    "        plt.plot(sigmas, ext_phen[i_gene, :],\n",
    "                 linestyle=linestyle,\n",
    "                 color=colors[i_gene],\n",
    "                 label=f'Gene {i_gene}')\n",
    "        \n",
    "    # Plot what an isolated gene looks like\n",
    "    final_exprs = 1.0 / (1.0 + np.exp((indiv.sigma_basal + sigmas - indiv.sigma_opt)/indiv.epsilon))\n",
    "\n",
    "    plt.plot(sigmas, final_exprs, color='tab:red')\n",
    "        \n",
    "    #plt.legend(ncol=1)\n",
    "                \n",
    "    if plot_name:\n",
    "        plt.savefig(plot_name + '.pdf', dpi=dpi, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0ff015f",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_activity_sigma_all_genes(init_indiv, plot_title='Random individual', plot_name='activity_random')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74c6318d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_all_single_switches(indiv):\n",
    "    for i_gene in range(indiv.nb_genes):\n",
    "        new_indiv = indiv.clone()\n",
    "        new_indiv.already_evaluated = False\n",
    "        # Switch one gene\n",
    "        new_indiv.genes[i_gene].orientation = 1 - new_indiv.genes[i_gene].orientation\n",
    "        plot_activity_sigma_all_genes(new_indiv,\n",
    "                                      plot_title=f'Random individual switched gene {i_gene}',\n",
    "                                      plot_name=f'activity_random_switch_{i_gene}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69b0786c",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_all_single_switches(init_indiv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a303171d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_phenotypic_distance(indiv):\n",
    "\n",
    "    nb_sigmas = 50\n",
    "    sigmas = np.linspace(-0.25, 0.15, nb_sigmas)\n",
    "\n",
    "    indiv_phen = compute_extended_phenotype(indiv, sigmas)\n",
    "\n",
    "    max_switches = 3\n",
    "\n",
    "    all_switches = []\n",
    "\n",
    "    # The distance of each individual with k gene switches from the original individual\n",
    "    for nb_switches in range(1, max_switches+1):\n",
    "        phen_dist = []\n",
    "\n",
    "        for genes_to_switch in itertools.combinations(range(indiv.nb_genes), nb_switches):\n",
    "            new_indiv = indiv.clone()\n",
    "            new_indiv.already_evaluated = False\n",
    "\n",
    "            # Switch genes\n",
    "            for i_gene in genes_to_switch: \n",
    "                new_indiv.genes[i_gene].orientation = 1 - new_indiv.genes[i_gene].orientation\n",
    "\n",
    "            new_indiv_phen = compute_extended_phenotype(new_indiv, sigmas)\n",
    "\n",
    "            phen_dist.append(np.sqrt(np.sum(np.square(indiv_phen - new_indiv_phen))))\n",
    "\n",
    "        all_switches.append(phen_dist)\n",
    "    \n",
    "    plt.xlabel('Number of switches')\n",
    "    plt.xticks(range(1, max_switches+1))\n",
    "    plt.ylabel('L2 distance')\n",
    "\n",
    "    plt.violinplot(all_switches, showmeans=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fdeb15a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_phenotypic_distance(init_indiv)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "518923a8",
   "metadata": {},
   "source": [
    "# Evolved individual: influence of env. supercoiling on final gene expression levels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ad47251",
   "metadata": {},
   "outputs": [],
   "source": [
    "# See how gene activity levels depend on environmental supercoiling\n",
    "def plot_activity_sigma_per_type(indiv, plot_all_genes=False, plot_title=None, plot_name=None):\n",
    "\n",
    "    colors = ['tab:blue', 'tab:red', 'tab:green'] # AB: blue, A: red, B: green\n",
    "    all_colors = mpl.cm.get_cmap('viridis', indiv.nb_genes)(range(indiv.nb_genes))\n",
    "\n",
    "    # Initialize the individual\n",
    "    indiv.evaluate(0.0, 0.0)\n",
    "\n",
    "    nb_sigmas = 250\n",
    "\n",
    "    activ = np.zeros((3, nb_sigmas)) # Compute activity for each gene type\n",
    "    sigmas = np.linspace(-0.2, 0.2, nb_sigmas)\n",
    "\n",
    "    for i_sigma, sigma_env in enumerate(sigmas):\n",
    "        # Evaluate the individual in the environment with sigma\n",
    "        temporal_expr = indiv.run_system(sigma_env)\n",
    "\n",
    "        # Compute total gene activation levels        \n",
    "        for i_gene, gene in enumerate(indiv.genes):\n",
    "            activ[gene.gene_type][i_sigma] += temporal_expr[i_gene, -1]\n",
    "            \n",
    "    activ /= (indiv.nb_genes / 3)\n",
    "            \n",
    "    plt.figure(figsize=(6, 4), dpi=dpi)\n",
    "    plt.xlabel('Environment supercoiling $\\sigma_{env}$')\n",
    "    plt.ylabel('Average gene activity by type')\n",
    "    plt.ylim(-0.05, 1.05)\n",
    "    plt.grid(linestyle=':')\n",
    "    \n",
    "    if plot_title:\n",
    "        plt.title(plot_title)\n",
    "\n",
    "    for i_gene_type, gene_type in enumerate(gene_types):\n",
    "        plt.plot(sigmas, activ[i_gene_type, :],\n",
    "                 color=gene_type_color[i_gene_type],\n",
    "                 linewidth=2,\n",
    "                 label=gene_type)\n",
    "\n",
    "    # Add sigma_A and sigma_B\n",
    "    y_min, y_max = plt.ylim()\n",
    "    plt.vlines(params['sigma_A'], y_min, y_max, linestyle='--', linewidth=2, color='black')\n",
    "    plt.vlines(params['sigma_B'], y_min, y_max, linestyle='--', linewidth=2, color='black')\n",
    "    plt.ylim(y_min, y_max)\n",
    "    \n",
    "    plt.legend()\n",
    "    \n",
    "    plt.tight_layout()\n",
    "        \n",
    "    if plot_name:\n",
    "        plt.savefig(plot_name, dpi=dpi, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f5701dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i_rep in range(15):\n",
    "    indiv = get_best_indiv(exp_path.joinpath(f'rep{i_rep:02}'), gen=200_000)\n",
    "    plot_activity_sigma_per_type(indiv, plot_title=f'Best replicate {i_rep}', plot_name=exp_path.joinpath(f'activity_best_rep{i_rep}.pdf'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7da4c03b",
   "metadata": {},
   "source": [
    "# Dispersion of gene activity levels with mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e87c5a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate N mutants from an individual to see dispersion of gene activity levels\n",
    "def plot_activity_mutation(indiv, sigma, mutation, plot_name=None):\n",
    "    \n",
    "    nb_mut = 1000\n",
    "\n",
    "    activ = np.zeros(nb_mut)\n",
    "\n",
    "    rng = np.random.default_rng()\n",
    "    \n",
    "    gene_positions, genome_size = indiv.compute_gene_positions(include_coding=False)\n",
    "\n",
    "    for i_mut in range(nb_mut):\n",
    "\n",
    "        # Generate a new mutant and evaluate it\n",
    "        mut_indiv = indiv.clone()\n",
    "        \n",
    "        start_pos = rng.integers(0, genome_size)\n",
    "        end_pos = rng.integers(0, genome_size)\n",
    "        \n",
    "        if end_pos < start_pos:\n",
    "            start_pos, end_pos = end_pos, start_pos\n",
    "\n",
    "        mut_indiv.perform_inversion(start_pos, end_pos)\n",
    "        mut_indiv.already_evaluated = False\n",
    "        \n",
    "        (temporal_expr, _), _ = mut_indiv.evaluate(sigma, sigma)\n",
    "\n",
    "        # Compute total gene activation levels\n",
    "        activ[i_mut] = np.sum(np.square(temporal_expr[:, -1])) /  mut_indiv.nb_genes\n",
    "\n",
    "    # Plot setup    \n",
    "    plt.figure(figsize=(9, 3), dpi=dpi)\n",
    "    plt.xlabel('Average squared gene activity')\n",
    "    #plt.xlim(-0.025, 0.525)\n",
    "    plt.ylabel('Number of mutants')\n",
    "    plt.grid(linestyle=':')\n",
    "    \n",
    "    # Plot the histogram\n",
    "    plt.hist(activ)\n",
    "    \n",
    "    # Plot the original activity\n",
    "    (orig_expr, _), _ = indiv.evaluate(sigma, sigma)\n",
    "    orig_activ = np.sum(np.square(temporal_expr[:, -1])) / indiv.nb_genes\n",
    "    y_min, y_max = plt.ylim()\n",
    "    plt.vlines(orig_activ, y_min, y_max, linestyle='--', linewidth=1,\n",
    "                   color='tab:red', label='Original activity level')\n",
    "    plt.ylim(y_min, y_max)\n",
    "    \n",
    "    plt.legend()\n",
    "\n",
    "    plt.tight_layout()\n",
    "    \n",
    "    if plot_name:\n",
    "        plt.savefig(plot_name + '.pdf', dpi=dpi, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28335aa8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "for i in range(5):\n",
    "    plot_activity_mutation(make_random_indiv(nb_mut=100), sigma=sigma_A, mutation=mutation, plot_name=f'robustness_random_{i}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6f13cad",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_activity_mutation(get_best_indiv(exp_path.joinpath('rep0'), gen=200_000), sigma=sigma_A, mutation=mutation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae3fe543",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_activity_mutation(get_best_indiv(exp_path.joinpath('rep0'), gen=200_000), sigma=sigma_B, mutation=mutation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "079efa71",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_activity_mutation(get_best_indiv(exp_path.joinpath('rep0'), gen=200_000), sigma=sigma_B, mutation=mutation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37bdf564",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "4af4ad38b24e346bac4a01c43b89b341085d1000b7b93cf2ecd0ab0adb9a79f7"
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
