{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pathlib\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy import stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import autoreload\n",
    "import evotsc_plot\n",
    "autoreload.reload(evotsc_plot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "label_fontsize=20\n",
    "tick_fontsize=15\n",
    "legend_fontsize=15\n",
    "dpi=300"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "exp_path = pathlib.Path('/Users/theotime/Desktop/evotsc/euler/substep_2_delta_1.0/')\n",
    "gen=200_000\n",
    "nb_genes_type = 20\n",
    "gene_types = ['AB', 'A', 'B'] # Name of each gene type\n",
    "gene_type_color = ['tab:blue', 'tab:red', 'tab:green'] #AB, A, B\n",
    "orient_name = ['leading', 'lagging'] # Name of each gene orientation\n",
    "rel_orient = ['conv', 'div', 'upstr', 'downstr']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_params(exp_path):\n",
    "    rep_dirs = sorted([d for d in exp_path.iterdir() if (d.is_dir() and d.name.startswith(\"rep\"))])\n",
    "    \n",
    "    with open(rep_dirs[0].joinpath('params.txt'), 'r') as params_file:\n",
    "        param_lines = params_file.readlines()\n",
    "        \n",
    "    params = {}\n",
    "    for line in param_lines:\n",
    "        param_name = line.split(':')[0]\n",
    "        if param_name == 'commit':\n",
    "            param_val = line.split(':')[1].strip()\n",
    "        else:\n",
    "            param_val = float(line.split(':')[1])\n",
    "        \n",
    "        params[param_name] = param_val\n",
    "        \n",
    "    return params"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "params = get_params(exp_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "params"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_stats(exp_name):\n",
    "    \n",
    "    exp_name = pathlib.Path(exp_name)\n",
    "        \n",
    "    rep_dirs = sorted([d for d in exp_name.iterdir() if (d.is_dir() and d.name.startswith(\"rep\"))])\n",
    "    \n",
    "    res = pd.DataFrame()\n",
    "    \n",
    "    for i_rep, rep_dir in enumerate(rep_dirs):\n",
    "                \n",
    "        res_dir = pd.read_csv(rep_dir.joinpath('stats.csv'))\n",
    "        \n",
    "        res_dir.insert(0, 'Replicate', i_rep)\n",
    "        \n",
    "        res = pd.concat([res, res_dir])\n",
    "    \n",
    "    return res"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot the number of active genes of each type over evolutionary time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_gene_activity(base_path, plot_sc=False, plot_genome_size=False):\n",
    "    \n",
    "    base_path = pathlib.Path(base_path)\n",
    "    \n",
    "    rep_dirs = sorted([d for d in exp_path.iterdir() if (d.is_dir() and d.name.startswith(\"rep\"))])\n",
    "    \n",
    "    nb_rep = len(rep_dirs)\n",
    "    \n",
    "    for i_rep, rep_dir in enumerate(rep_dirs):\n",
    "        stats_path = rep_dir.joinpath('stats.csv')\n",
    "        if stats_path.stat().st_size > 0: # not an empty file\n",
    "\n",
    "            data = pd.read_csv(stats_path)\n",
    "            data = data[data['Gen'] > 0]\n",
    "\n",
    "            for env in [\"A\", \"B\"]:\n",
    "\n",
    "                fig, ax1 = plt.subplots(figsize=(9, 4), dpi=dpi)\n",
    "                ax1.set_ylim(-0.05, 1.05)\n",
    "                ax1.set_ylabel('Proportion of activated genes', fontsize=label_fontsize)\n",
    "                ax1.set_xlabel('Generation', fontsize=label_fontsize)\n",
    "                ax1.set_xscale('log')\n",
    "                ax1.grid()\n",
    "\n",
    "                ax1.plot(data['Gen'], data[f\"ABon_{env}\"] / (data[f\"ABon_{env}\"] + data[f\"ABoff_{env}\"]),\n",
    "                         color=\"tab:blue\",\n",
    "                         linewidth=2,\n",
    "                         label=\"AB genes on\")\n",
    "                ax1.plot(data['Gen'], data[f\"Aon_{env}\"] / (data[f\"Aon_{env}\"] + data[f\"Aoff_{env}\"]),\n",
    "                         color=\"tab:red\",\n",
    "                         linewidth=2,\n",
    "                         label=\"A genes on\")\n",
    "                ax1.plot(data['Gen'], data[f\"Bon_{env}\"] / (data[f\"Bon_{env}\"] + data[f\"Boff_{env}\"]),\n",
    "                         color=\"tab:green\",\n",
    "                         linewidth=2,\n",
    "                         label=\"B genes on\")\n",
    "                ax1.tick_params(axis='both', which='major', labelsize=tick_fontsize)\n",
    "\n",
    "\n",
    "                ## 2nd axis: fitness\n",
    "                ax2 = ax1.twinx()\n",
    "                ax2.set_yscale('log')\n",
    "                ax2.set_ylim(1e-25, 1e0)\n",
    "                ax2.set_ylabel('Fitness', fontsize=label_fontsize, color='tab:cyan')\n",
    "                ax2.plot(data['Gen'],\n",
    "                         data[\"Fitness\"],\n",
    "                         color=\"tab:cyan\",\n",
    "                         linewidth=2)\n",
    "                ax2.tick_params(axis='both', which='major', labelsize=tick_fontsize)\n",
    "\n",
    "                \n",
    "                ## 3rd axis: supercoiling\n",
    "                if plot_genome_size:\n",
    "                    ax3 = ax1.twinx()\n",
    "\n",
    "                    ax3.spines['right'].set_position(('outward', 80))\n",
    "                    ax3.set_ylim(0.7e4, 6.3e4)\n",
    "                    ax3.set_ylabel('Genome Size', fontsize=label_fontsize, color='tab:olive')\n",
    "                    ax3.plot(data['Gen'],\n",
    "                             data[\"Genome size\"],\n",
    "                             color=\"tab:olive\",\n",
    "                             linewidth=2)\n",
    "                    ax3.tick_params(axis='both', which='major', labelsize=tick_fontsize)\n",
    "                \n",
    "                #plt.title(f\"Environment {env}\")\n",
    "                fig.legend(bbox_to_anchor=(0,0),\n",
    "                           bbox_transform=ax1.transAxes,\n",
    "                           loc=\"lower left\",\n",
    "                           fontsize=legend_fontsize)            \n",
    "\n",
    "                rep_num = rep_dir.name[3:] # add a `_` between the `rep` and the rep number \n",
    "                plt.savefig(f'{base_path}/rep_{rep_num}_env_{env}.pdf', dpi=dpi, bbox_inches='tight')\n",
    "                \n",
    "                plt.close('all')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_gene_activity(exp_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot the mean and stddev of activated genes of each type in each environment over all replicas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "full_stats = get_stats(exp_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_mean_std(full_stats):\n",
    "    last_gen = np.min(full_stats.groupby('Replicate').max()['Gen'])\n",
    "    last_gen_stats = full_stats[full_stats['Gen'] == last_gen] / nb_genes_type\n",
    "    mean_stats = last_gen_stats.mean()\n",
    "    std_stats = last_gen_stats.std()\n",
    "    fig, ax = plt.subplots(figsize=(9, 4), dpi=dpi)\n",
    "\n",
    "    x = np.arange(3)  # 3 types of genes\n",
    "    width = 0.35  # the width of the bars\n",
    "\n",
    "    env_A_means = np.array([mean_stats[\"ABon_A\"], mean_stats[\"Aon_A\"], mean_stats[\"Bon_A\"]])\n",
    "    env_B_means = np.array([mean_stats[\"ABon_B\"], mean_stats[\"Aon_B\"], mean_stats[\"Bon_B\"]])\n",
    "\n",
    "    env_A_std = np.array([std_stats[\"ABon_A\"], std_stats[\"Aon_A\"], std_stats[\"Bon_A\"]])\n",
    "    env_B_std = np.array([std_stats[\"ABon_B\"], std_stats[\"Aon_B\"], std_stats[\"Bon_B\"]])\n",
    "\n",
    "\n",
    "    rects_A = ax.bar(x - width/2,\n",
    "                    env_A_means,\n",
    "                    width=width,\n",
    "                    label='Environment A',\n",
    "                    #yerr=[np.zeros_like(env_A_std), env_A_std],\n",
    "                    yerr=env_A_std,\n",
    "                    capsize=5)\n",
    "\n",
    "    rects_B = ax.bar(x + width/2,\n",
    "                     env_B_means,\n",
    "                     width=width,\n",
    "                     label='Environment B',\n",
    "                     yerr=env_B_std,\n",
    "                     capsize=5)\n",
    "\n",
    "    ax.set_ylabel('Activated genes', fontsize=label_fontsize)\n",
    "    ax.set_xticks(x)\n",
    "    ax.set_xticklabels([\"AB genes on\", \"A genes on\", \"B genes on\"])\n",
    "\n",
    "    plt.grid(linestyle=':', axis='y')\n",
    "\n",
    "    plt.tick_params(axis='both', which='major', labelsize=tick_fontsize)\n",
    "\n",
    "    plt.legend(fontsize=legend_fontsize, loc='upper center')\n",
    "\n",
    "    plt.savefig(f'{exp_path}/mean_activation.pdf', bbox_inches='tight')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_mean_std(full_stats)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Statistical significance tests for the above figure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def stats_tests(full_stats):\n",
    "    last_gen = np.min(full_stats.groupby('Replicate').max()['Gen'])\n",
    "    last_gen_stats = full_stats[full_stats[\"Gen\"] == last_gen] / nb_genes_type\n",
    "    stats_AB = stats.ttest_rel(last_gen_stats[\"ABon_A\"], last_gen_stats[\"ABon_B\"])\n",
    "    stats_Aon = stats.ttest_rel(last_gen_stats[\"Aon_A\"], last_gen_stats[\"Aon_B\"])\n",
    "    stats_Bon = stats.ttest_rel(last_gen_stats[\"Bon_A\"], last_gen_stats[\"Bon_B\"])\n",
    "    print(f'AB genes: {stats_AB}')\n",
    "    print(f'A genes on: {stats_Aon}')\n",
    "    print(f'B genes on: {stats_Bon}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "stats_tests(full_stats)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot fitness, genome size, and basal supercoiling over evolutionary time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_fitness(full_stats):\n",
    "    nb_rep = full_stats[\"Replicate\"].nunique()\n",
    "    \n",
    "    colors = mpl.cm.get_cmap('viridis', nb_rep)(range(nb_rep))\n",
    "    \n",
    "    plt.figure(figsize=(9,4), dpi=dpi)\n",
    "    \n",
    "    plt.xscale('log')\n",
    "    plt.yscale('log')\n",
    "    plt.grid(linestyle=':')\n",
    "    \n",
    "    plt.xlabel('Generation', fontsize=label_fontsize)\n",
    "    plt.ylabel('Fitness', fontsize=label_fontsize)\n",
    "    \n",
    "    plt.tick_params(axis='both', which='major', labelsize=tick_fontsize)\n",
    "    \n",
    "    for rep in range(nb_rep):\n",
    "        stats_rep = full_stats[full_stats[\"Replicate\"] == rep]\n",
    "        stats_rep = stats_rep[stats_rep[\"Gen\"] > 0]\n",
    "        plt.plot(stats_rep['Gen'],\n",
    "                 stats_rep[\"Fitness\"],\n",
    "                 linewidth=2,\n",
    "                 color=colors[rep])\n",
    "        \n",
    "    plt.savefig(f'{exp_path}/all_fitness.pdf', dpi=dpi, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_fitness(full_stats)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_sigma(full_stats):\n",
    "    \n",
    "    if 'basal_sc' not in full_stats.columns:\n",
    "        return\n",
    "    \n",
    "    nb_rep = full_stats[\"Replicate\"].nunique()\n",
    "    \n",
    "    colors = mpl.cm.get_cmap('viridis', nb_rep)(range(nb_rep))\n",
    "    \n",
    "    fig, ax1 = plt.subplots(figsize=(9,4), dpi=dpi)\n",
    "    \n",
    "    #plt.xscale('log')\n",
    "    #plt.yscale('log')\n",
    "    plt.grid(linestyle=':')\n",
    "    \n",
    "    ax1.set_xlabel('Generation', fontsize=label_fontsize)\n",
    "    ax1.set_ylabel('Basal SC', fontsize=label_fontsize)\n",
    "    \n",
    "    plt.tick_params(axis='both', which='major', labelsize=tick_fontsize)\n",
    "        \n",
    "    for rep in range(nb_rep):\n",
    "        stats_rep = full_stats[full_stats[\"Replicate\"] == rep]\n",
    "        stats_rep = stats_rep[stats_rep[\"Gen\"] > 0]\n",
    "        ax1.plot(stats_rep[\"Gen\"],\n",
    "                 stats_rep[\"basal_sc\"],\n",
    "                 linewidth=2,\n",
    "                 color=colors[rep])\n",
    "        \n",
    "    plt.savefig(f'{exp_path}/all_basal_sc.pdf', dpi=dpi, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plot_sigma(full_stats)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_genome_size(full_stats):\n",
    "    \n",
    "    if 'Genome size' not in full_stats.columns:\n",
    "        return\n",
    "    \n",
    "    nb_rep = full_stats[\"Replicate\"].nunique()\n",
    "    \n",
    "    colors = mpl.cm.get_cmap('viridis', nb_rep)(range(nb_rep))\n",
    "    \n",
    "    fig, ax1 = plt.subplots(figsize=(9,4), dpi=dpi)\n",
    "    \n",
    "    #plt.xscale('log')\n",
    "    #plt.yscale('log')\n",
    "    plt.grid(linestyle=':')\n",
    "    \n",
    "    ax1.set_xlabel('Generation', fontsize=label_fontsize)\n",
    "    ax1.set_ylabel('Genome Size', fontsize=label_fontsize)\n",
    "    \n",
    "    #ax1.set_ylim(0, 189000)\n",
    "    \n",
    "    plt.tick_params(axis='both', which='major', labelsize=tick_fontsize)\n",
    "        \n",
    "    for rep in range(nb_rep):\n",
    "        stats_rep = full_stats[full_stats[\"Replicate\"] == rep]\n",
    "        stats_rep = stats_rep[stats_rep[\"Gen\"] > 0]\n",
    "        ax1.plot(stats_rep[\"Gen\"],\n",
    "                 stats_rep[\"Genome size\"],\n",
    "                 linewidth=2,\n",
    "                 color=colors[rep])\n",
    "        \n",
    "    plt.savefig(f'{exp_path}/all_genome_size.pdf', dpi=dpi, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_genome_size(full_stats)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot gene activities over the lifecycle of the best individual in each replica"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_expr_AB(indiv, sigma_A, sigma_B, plot_title, plot_name):\n",
    "\n",
    "    (temporal_expr_A, temporal_expr_B), fitness = indiv.evaluate(sigma_A, sigma_B)\n",
    "\n",
    "    colors = ['tab:blue', 'tab:red', 'tab:green'] # AB: blue, A: red, B: green\n",
    "\n",
    "    plt.figure(figsize=(9, 8), dpi=dpi)\n",
    "\n",
    "    ## First subplot: environment A\n",
    "    plt.subplot(2, 1, 1)\n",
    "    plt.ylim(-0.05, 1.05)\n",
    "\n",
    "    for i_gene, gene in enumerate(indiv.genes):\n",
    "        linestyle = 'solid' if gene.orientation == 0 else 'dashed'\n",
    "        plt.plot(temporal_expr_A[i_gene, :],\n",
    "                 linestyle=linestyle,\n",
    "                 linewidth=2,\n",
    "                 color=colors[gene.gene_type],\n",
    "                 #alpha=0.25,\n",
    "                 label=f'Gene {gene.id}')\n",
    "\n",
    "    plt.grid(linestyle=':')\n",
    "    #plt.xlabel('Time', fontsize='large')\n",
    "    plt.ylabel('Expression level', fontsize=label_fontsize)\n",
    "\n",
    "    plt.tick_params(axis='both', which='major', labelsize=tick_fontsize)\n",
    "\n",
    "    #plt.legend(loc='center right')\n",
    "    #plt.title('Environment A')\n",
    "\n",
    "    ## Second subplot: environment B\n",
    "    plt.subplot(2, 1, 2)\n",
    "    plt.ylim(-0.05, 1.05)\n",
    "\n",
    "    for i_gene, gene in enumerate(indiv.genes):\n",
    "        linestyle = 'solid' if gene.orientation == 0 else 'dashed'\n",
    "        plt.plot(temporal_expr_B[i_gene, :],\n",
    "                 linestyle=linestyle,\n",
    "                 linewidth=2,\n",
    "                 color=colors[gene.gene_type],\n",
    "                 #alpha=0.25,\n",
    "                 label=f'Gene {gene.id}')\n",
    "\n",
    "    plt.grid(linestyle=':')\n",
    "    plt.xlabel('Time', fontsize=label_fontsize)\n",
    "    plt.ylabel('Expression level', fontsize=label_fontsize)\n",
    "\n",
    "    plt.tick_params(axis='both', which='major', labelsize=tick_fontsize)\n",
    "\n",
    "    #plt.legend(loc='center right')\n",
    "    #plt.title('Environment B')\n",
    "\n",
    "    ## Final stuff\n",
    "\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(plot_name + '.pdf', dpi=dpi, bbox_inches='tight')\n",
    "    plt.show()\n",
    "    plt.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_best_indiv(rep_path, gen):\n",
    "    \n",
    "    with open(rep_path.joinpath(f'pop_gen_{gen:06}.evotsc'), 'rb') as save_file:\n",
    "        pop_rep = pickle.load(save_file)\n",
    "        \n",
    "    pop_rep.evaluate()\n",
    "    \n",
    "    best_fit = 0\n",
    "    best_indiv = pop_rep.individuals[0]\n",
    "    \n",
    "    try:\n",
    "        for indiv in pop_rep.individuals:\n",
    "            if indiv.fitness > best_fit:\n",
    "                best_fit = indiv.fitness\n",
    "                best_indiv = indiv\n",
    "    except AttributeError: # In the neutral control, individuals are not evaluated so there is no fitness field\n",
    "        pass\n",
    "    \n",
    "    return best_indiv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_best(gen):\n",
    "    rep_dirs = sorted([d for d in exp_path.iterdir() if (d.is_dir() and d.name.startswith(\"rep\"))])\n",
    "\n",
    "    for rep, rep_dir in enumerate(rep_dirs):\n",
    "        best_rep = get_best_indiv(rep_dir, gen)\n",
    "        plot_expr_AB(best_rep, sigma_A=params['sigma_A'], sigma_B=params['sigma_B'],\n",
    "                     plot_title='', plot_name=f'{exp_path}/best_rep{rep}')\n",
    "        #evotsc_plot.plot_genome(best_rep, name=f'{exp_path}/genome_rep{rep}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_best(gen)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot average gene expression levels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_avg_expr(indiv, sigma_env, plot_title, plot_name):\n",
    "\n",
    "    temporal_expr = indiv.run_system(sigma_env)\n",
    "\n",
    "    nb_genes, nb_steps = temporal_expr.shape\n",
    "\n",
    "    colors = mpl.cm.get_cmap('viridis', nb_genes)(range(nb_genes))\n",
    "\n",
    "    plt.figure(figsize=(9, 4), dpi=dpi)\n",
    "\n",
    "    plt.ylim(-0.05, 1.05)\n",
    "    \n",
    "    type_expr = np.zeros((3, nb_steps))\n",
    "    \n",
    "    for wanted_gene_type in gene_types:\n",
    "        for i_gene, gene in enumerate(indiv.genes):\n",
    "            if gene_types[gene.gene_type] == wanted_gene_type:\n",
    "                type_expr[gene.gene_type] += temporal_expr[i_gene, :]\n",
    "        \n",
    "    type_expr /= (nb_genes // 3)\n",
    "\n",
    "    for i_gene_type, gene_type in enumerate(gene_types):\n",
    "        plt.plot(type_expr[i_gene_type, :],\n",
    "                 color=gene_type_color[i_gene_type],\n",
    "                 label=f'{gene_type} genes')\n",
    "\n",
    "    plt.grid(linestyle=':')\n",
    "    plt.xlabel('Time', fontsize='large')\n",
    "    plt.ylabel('Expression level', fontsize='large')\n",
    "    \n",
    "    x_min, x_max = plt.xlim()\n",
    "    plt.hlines(0.5, x_min, x_max, linestyle='--', linewidth=1,\n",
    "           color='tab:red', label='Half activation threshold')\n",
    "    plt.xlim(x_min, x_max)\n",
    "\n",
    "\n",
    "    plt.legend(loc='lower left')\n",
    "    plt.title(plot_title)\n",
    "    \n",
    "    plt.savefig(exp_path.joinpath(plot_name + '.pdf'), dpi=300, bbox_inches='tight')\n",
    "\n",
    "    plt.show()\n",
    "\n",
    "    plt.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_avg_expr_all(exp_path, gen):\n",
    "    rep_dirs = sorted([d for d in exp_path.iterdir() if (d.is_dir() and d.name.startswith(\"rep\"))])\n",
    "    \n",
    "    for rep, rep_dir in enumerate(rep_dirs):\n",
    "        \n",
    "        indiv = get_best_indiv(rep_dir, gen)\n",
    "        \n",
    "        plot_avg_expr(indiv, params['sigma_A'], 'Environment A', f'avg_expr_rep{rep:02}_env_A')\n",
    "        #plot_avg_expr(indiv, params['sigma_B'], 'Environment B', f'avg_expr_rep{rep:02}_env_B')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plot_avg_expr_all(exp_path, gen)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot average supercoiling for each type of gene over replicas (_il faut changer les moyennes en médianes_)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sigma_tsc(indiv, sigma_env: float):\n",
    "    temporal_expr = np.zeros((indiv.nb_genes, indiv.nb_eval_steps))\n",
    "\n",
    "    # Initial values at t = 0\n",
    "    temporal_expr[:, 0] = np.array([gene.basal_expression for gene in indiv.genes])\n",
    "    \n",
    "    sigma_tsc = np.zeros((indiv.nb_genes, indiv.nb_eval_steps))\n",
    "\n",
    "    # Iterate the system\n",
    "    for t in range(1, indiv.nb_eval_steps):\n",
    "        sigma_local = indiv.inter_matrix @ temporal_expr[:, t-1]\n",
    "        sigma_tsc[:, t] = sigma_local\n",
    "        sigma_total = indiv.sigma_basal + sigma_local + sigma_env\n",
    "        temporal_expr[:, t] = 1.0 / (1.0 + np.exp((sigma_total - indiv.sigma_opt)/indiv.epsilon))\n",
    "\n",
    "    return sigma_tsc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_avg_sigma(exp_path, gen):\n",
    "    \n",
    "    rep_dirs = sorted([d for d in exp_path.iterdir() if (d.is_dir() and d.name.startswith(\"rep\"))])\n",
    "    nb_reps = len(rep_dirs)\n",
    "    \n",
    "    time_steps = 5\n",
    "    \n",
    "    full_data = pd.DataFrame()\n",
    "\n",
    "    for rep, rep_dir in enumerate(rep_dirs):\n",
    "        best_indiv = get_best_indiv(rep_dir, gen)\n",
    "        \n",
    "        sigma_A = get_sigma_tsc(best_indiv, sigma_env=params['sigma_A'])\n",
    "        sigma_B = get_sigma_tsc(best_indiv, sigma_env=params['sigma_B'])\n",
    "\n",
    "        # Take temporal averages\n",
    "        mean_sigma_A_gene = np.sum(sigma_A[:, best_indiv.nb_eval_steps-time_steps:], axis=1) / time_steps\n",
    "        mean_sigma_B_gene = np.sum(sigma_B[:, best_indiv.nb_eval_steps-time_steps:], axis=1) / time_steps\n",
    "        \n",
    "        # Sort by gene type\n",
    "        mean_sigma_A = np.zeros(3)\n",
    "        mean_sigma_B = np.zeros(3)\n",
    "        for i_gene in range(best_indiv.nb_genes):\n",
    "            mean_sigma_A[best_indiv.genes[i_gene].gene_type] += mean_sigma_A_gene[i_gene]\n",
    "            mean_sigma_B[best_indiv.genes[i_gene].gene_type] += mean_sigma_B_gene[i_gene]\n",
    "\n",
    "        # Divide by number of genes in each class\n",
    "        mean_sigma_A /= (best_indiv.nb_genes // 3)\n",
    "        mean_sigma_B /= (best_indiv.nb_genes // 3)\n",
    "        \n",
    "        # Add other supercoiling sources\n",
    "        mean_sigma_A += best_indiv.sigma_basal + params['sigma_A']\n",
    "        mean_sigma_B += best_indiv.sigma_basal + params['sigma_B']\n",
    "        \n",
    "        col_names = []\n",
    "        \n",
    "        for gene_type in gene_types:\n",
    "            col_names.append(f'{gene_type}sc_A')\n",
    "            \n",
    "        for gene_type in gene_types:\n",
    "            col_names.append(f'{gene_type}sc_B')\n",
    "                    \n",
    "        rep_data = pd.DataFrame(data=[np.concatenate([mean_sigma_A, mean_sigma_B])], \n",
    "                                columns=col_names)\n",
    "        \n",
    "        rep_data.insert(0, 'Replicate', rep)\n",
    "        \n",
    "        full_data = pd.concat([full_data, rep_data])\n",
    "\n",
    "    full_data = full_data.set_index('Replicate')\n",
    "\n",
    "    return full_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_mean_sigma(exp_path, gen, title=None):\n",
    "    \n",
    "    full_data = get_avg_sigma(exp_path, gen)\n",
    "                \n",
    "    ## Actual plotting\n",
    "    mean_stats = full_data.mean()\n",
    "    std_stats = full_data.std()\n",
    "    med_stats = full_data.median()\n",
    "    \n",
    "    fig, ax = plt.subplots(figsize=(9, 4), dpi=300)\n",
    "\n",
    "    x = np.arange(3)  # 3 types of genes\n",
    "    width = 0.1  # the width of the bars\n",
    "\n",
    "    env_A_means = np.array([mean_stats[\"ABsc_A\"], mean_stats[\"Asc_A\"], mean_stats[\"Bsc_A\"]])\n",
    "    env_B_means = np.array([mean_stats[\"ABsc_B\"], mean_stats[\"Asc_B\"], mean_stats[\"Bsc_B\"]])\n",
    "\n",
    "    env_A_std = np.array([std_stats[\"ABsc_A\"], std_stats[\"Asc_A\"], std_stats[\"Bsc_A\"]])\n",
    "    env_B_std = np.array([std_stats[\"ABsc_B\"], std_stats[\"Asc_B\"], std_stats[\"Bsc_B\"]])\n",
    "    \n",
    "    env_A_med = np.array([med_stats[\"ABsc_A\"], med_stats[\"Asc_A\"], med_stats[\"Bsc_A\"]])\n",
    "    env_B_med = np.array([med_stats[\"ABsc_B\"], med_stats[\"Asc_B\"], med_stats[\"Bsc_B\"]])\n",
    "\n",
    "\n",
    "    # Plot mean and std\n",
    "    ax.errorbar(x - width/2,\n",
    "                env_A_means,\n",
    "                label='Environment A',\n",
    "                marker='o',\n",
    "                linestyle='',\n",
    "                color='tab:blue',\n",
    "                yerr=env_A_std,\n",
    "                capsize=5)\n",
    "\n",
    "    ax.errorbar(x + width/2,\n",
    "                env_B_means,\n",
    "                marker='o',\n",
    "                linestyle='',\n",
    "                color='tab:orange',\n",
    "                label='Environment B',\n",
    "                yerr=env_B_std,\n",
    "                capsize=5)\n",
    "    \n",
    "    # Add median values -------- note : c'est la médiane des *moyennes*, or ce sont les *médianes* qu'il faut\n",
    "    ax.plot(x - width/2,\n",
    "            env_A_med,\n",
    "            marker='x',\n",
    "            linestyle='',\n",
    "            color='tab:blue')\n",
    "    \n",
    "    ax.plot(x + width/2,\n",
    "            env_B_med,\n",
    "            marker='x',\n",
    "            linestyle='',\n",
    "            color='tab:orange')\n",
    "    \n",
    "    \n",
    "    # Add half-activation threshold (sigma_opt)\n",
    "    x_min, x_max = ax.get_xlim()\n",
    "    ax.hlines(params['sigma_opt'], x_min, x_max, linestyle='--', linewidth=1,\n",
    "           color='tab:red', label='Half activation threshold')\n",
    "    ax.set_xlim(x_min, x_max)\n",
    "\n",
    "    ax.set_ylabel('SC level', fontsize=label_fontsize)\n",
    "    ax.set_ylim(0.25, -0.85)\n",
    "    ax.set_xticks(x)\n",
    "    ax.set_xticklabels([\"AB genes\", \"A genes\", \"B genes\"])\n",
    "\n",
    "    plt.grid(linestyle=':', axis='y')\n",
    "\n",
    "    plt.tick_params(axis='both', which='major', labelsize=tick_fontsize)\n",
    "\n",
    "    plt.legend(fontsize=legend_fontsize, loc='upper center')\n",
    "    \n",
    "    if title:\n",
    "        plt.title(title, fontsize=label_fontsize)\n",
    "\n",
    "    plt.savefig(f'{exp_path}/mean_supercoiling.pdf', bbox_inches='tight')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_mean_sigma(exp_path, gen, title=exp_path.name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
