{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import datetime\n",
    "import pickle\n",
    "import pathlib"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import autoreload\n",
    "import evotsc\n",
    "import evotsc_plot\n",
    "autoreload.reload(evotsc)\n",
    "autoreload.reload(evotsc_plot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "exp_path = pathlib.Path('/Users/theotime/Desktop/evotsc/vanilla_200k/')\n",
    "neutral_exp_path = pathlib.Path('/Users/theotime/Desktop/evotsc/neutral_100k/')\n",
    "gen=200_000\n",
    "gene_types = ['AB', 'A', 'B'] # Name of each gene type\n",
    "gene_type_color = ['tab:blue', 'tab:red', 'tab:green'] #AB, A, B\n",
    "orient_name = ['leading', 'lagging'] # Name of each gene orientation\n",
    "rel_orient = ['conv', 'div', 'upstr', 'downstr']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "label_fontsize=20\n",
    "tick_fontsize=15\n",
    "legend_fontsize=15\n",
    "dpi=300"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_params(exp_path):\n",
    "    rep_dirs = sorted([d for d in exp_path.iterdir() if (d.is_dir() and d.name.startswith(\"rep\"))])\n",
    "    \n",
    "    with open(rep_dirs[0].joinpath('params.txt'), 'r') as params_file:\n",
    "        param_lines = params_file.readlines()\n",
    "        \n",
    "    params = {}\n",
    "    for line in param_lines:\n",
    "        param_name = line.split(':')[0]\n",
    "        if param_name == 'commit':\n",
    "            param_val = line.split(':')[1].strip()\n",
    "        else:\n",
    "            param_val = float(line.split(':')[1])\n",
    "        \n",
    "        params[param_name] = param_val\n",
    "        \n",
    "    return params"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "params = get_params(exp_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "params"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_best_indiv(rep_path, gen):\n",
    "    \n",
    "    with open(rep_path.joinpath(f'pop_gen_{gen}.evotsc'), 'rb') as save_file:\n",
    "        pop_rep = pickle.load(save_file)\n",
    "        \n",
    "    pop_rep.evaluate()\n",
    "    \n",
    "    best_fit = 0\n",
    "    best_indiv = pop_rep.individuals[0]\n",
    "    \n",
    "    try:\n",
    "        for indiv in pop_rep.individuals:\n",
    "            if indiv.fitness > best_fit:\n",
    "                best_fit = indiv.fitness\n",
    "                best_indiv = indiv\n",
    "    except AttributeError: # In the neutral control, individuals are not evaluated so there is no fitness field\n",
    "        pass\n",
    "    \n",
    "    return best_indiv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_best_genome(exp_path, gen):\n",
    "    rep_dirs = sorted([d for d in exp_path.iterdir() if (d.is_dir() and d.name.startswith(\"rep\"))])\n",
    "    \n",
    "    for rep, rep_dir in enumerate(rep_dirs):\n",
    "        \n",
    "        best_indiv = get_best_indiv(rep_dir, gen)\n",
    "\n",
    "        evotsc_plot.plot_genome(best_indiv, name=exp_path.joinpath(f'genome_rep{rep:02}.png'))\n",
    "        #evotsc_plot.plot_expr_AB(best_indiv,\n",
    "        #                         sigma_A=params['sigma_A'],\n",
    "        #                         sigma_B=params['sigma_B'],\n",
    "        #                         plot_title=f'best rep {rep:02}',\n",
    "        #                         plot_name=exp_path.joinpath(f'best_rep{rep:02}.png'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_best_genome(exp_path, gen)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot gene pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_gene_stats(indiv):\n",
    "    result_dict = {'gene_id': [],\n",
    "                   'gene_type': [],\n",
    "                   'gene_orient': [],\n",
    "                   'other_id': [],\n",
    "                   'other_type': [],\n",
    "                   'other_orient': [],\n",
    "                   'rel_orient': [],\n",
    "                   'distance': []}\n",
    "    \n",
    "    for i_gene, gene in enumerate(indiv.genes):\n",
    "        #                                                      other - gene\n",
    "        other = indiv.genes[i_gene - 1]\n",
    "        result_dict['gene_id'].append(gene.id)\n",
    "        result_dict['gene_type'].append(gene_types[gene.gene_type])\n",
    "        result_dict['gene_orient'].append(orient_name[gene.orientation])\n",
    "        result_dict['other_id'].append(other.id)\n",
    "        result_dict['other_type'].append(gene_types[other.gene_type])\n",
    "        result_dict['other_orient'].append(orient_name[other.orientation])\n",
    "        if other.orientation == 0 and gene.orientation == 1:   # ---> <---\n",
    "            rel_orient = 'conv'\n",
    "        elif other.orientation == 0 and gene.orientation == 0: # ---> --->\n",
    "            rel_orient = 'downstr'\n",
    "        elif other.orientation == 1 and gene.orientation == 0: # <--- --->\n",
    "            rel_orient = 'div'\n",
    "        else:                                                  # <--- <---\n",
    "            rel_orient = 'upstr'\n",
    "        result_dict['rel_orient'].append(rel_orient)\n",
    "        result_dict['distance'].append(other.intergene)\n",
    "\n",
    "        #                                                       gene - other\n",
    "        other = indiv.genes[(i_gene + 1) % indiv.nb_genes]\n",
    "        result_dict['gene_id'].append(gene.id)\n",
    "        result_dict['gene_type'].append(gene_types[gene.gene_type])\n",
    "        result_dict['gene_orient'].append(orient_name[gene.orientation])\n",
    "        result_dict['other_id'].append(other.id)\n",
    "        result_dict['other_type'].append(gene_types[other.gene_type])\n",
    "        result_dict['other_orient'].append(orient_name[other.orientation])\n",
    "        if gene.orientation == 0 and other.orientation == 1:   # ---> <---\n",
    "            rel_orient = 'conv'\n",
    "        elif gene.orientation == 0 and other.orientation == 0: # ---> --->\n",
    "            rel_orient = 'upstr'\n",
    "        elif gene.orientation == 1 and other.orientation == 0: # <--- --->\n",
    "            rel_orient = 'div'\n",
    "        else:                                                  # <--- <---\n",
    "            rel_orient = 'downstr'\n",
    "        result_dict['rel_orient'].append(rel_orient)\n",
    "        result_dict['distance'].append(gene.intergene)\n",
    "\n",
    "    return pd.DataFrame.from_dict(result_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_full_stats(exp_path, gen):\n",
    "    rep_dirs = sorted([d for d in exp_path.iterdir() if (d.is_dir() and d.name.startswith(\"rep\"))])\n",
    "    \n",
    "    full_stats = pd.DataFrame()\n",
    "        \n",
    "    for rep, rep_dir in enumerate(rep_dirs):\n",
    "        \n",
    "        best_indiv = get_best_indiv(rep_dir, gen)\n",
    "\n",
    "        indiv_stats = get_gene_stats(best_indiv)\n",
    "        indiv_stats.insert(0, 'Replicate', rep)\n",
    "        full_stats = pd.concat([full_stats, indiv_stats])\n",
    "            \n",
    "    return full_stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "evol_stats = get_full_stats(exp_path, gen)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "neutral_stats = get_full_stats(neutral_exp_path, 100000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_gene_stats(stats, plot_name, count_bars):\n",
    "    \n",
    "    # If count_bars is true, the bars represent the count for each pair type and\n",
    "    # we annotate them with the average distances; if count_bars is false, we \n",
    "    # plot the opposite: the bars represent average distances, and are annotated\n",
    "    # with the count for each pair type.\n",
    "    \n",
    "    fig, axs = plt.subplots(3, 3, sharey='all', figsize=(12, 12), dpi=300)\n",
    "    \n",
    "    group_cols = ['gene_type', 'other_type', 'rel_orient']\n",
    "    \n",
    "    idx = pd.MultiIndex.from_product([gene_types, gene_types, rel_orient], names=group_cols)\n",
    "\n",
    "    grouped_stats = stats.groupby(group_cols).count().reindex(idx, fill_value=0).sort_index()\n",
    "    mean_stats = stats.groupby(group_cols).mean().reindex(idx, fill_value=0).sort_index()\n",
    "    std_stats = stats.groupby(group_cols).mean().reindex(idx, fill_value=0).sort_index()\n",
    "    \n",
    "    nb_reps = stats['Replicate'].nunique()\n",
    "    \n",
    "    colors = ['tab:blue', 'tab:orange', 'tab:green', 'tab:red']\n",
    "    \n",
    "    x = np.arange(len(rel_orient))\n",
    "\n",
    "    for i_gene_type, gene_type in enumerate(gene_types):\n",
    "        for i_other_type, other_type in enumerate(gene_types):\n",
    "            for cur_orient in enumerate(rel_orient):\n",
    "                \n",
    "                if count_bars:\n",
    "                    bar_data = grouped_stats.loc[(gene_type, other_type)]['Replicate'] / nb_reps\n",
    "                    text_data = mean_stats.loc[(gene_type, other_type)]['distance']\n",
    "                    \n",
    "                else:\n",
    "                    bar_data = mean_stats.loc[(gene_type, other_type)]['distance'] / nb_reps\n",
    "                    text_data = grouped_stats.loc[(gene_type, other_type)]['Replicate']\n",
    "                \n",
    "\n",
    "                ax = axs[i_gene_type][i_other_type]\n",
    "                \n",
    "                rects = ax.bar(x, bar_data, color=colors)\n",
    "                \n",
    "                if count_bars:\n",
    "                    ax.set_ylim(0, 20)\n",
    "                #    \n",
    "                #else:\n",
    "                #    ax.set_ylim(0, 3250)\n",
    "                \n",
    "                ax.set_xticks(x)\n",
    "                ax.set_xticklabels(labels=rel_orient, fontsize=\"large\")\n",
    "                \n",
    "                #Label rects with # of events\n",
    "                for i_rect, rect in enumerate(rects):\n",
    "                    ax.annotate(f\"{text_data[i_rect]:.1f}\",\n",
    "                                xy=(rect.get_x() + rect.get_width()/2, 0),\n",
    "                                xytext=(0, 3),\n",
    "                                ha='center',\n",
    "                                textcoords=\"offset points\",\n",
    "                                color='black')\n",
    "                \n",
    "                ax.grid(b=True, axis='y', linestyle=':')\n",
    "\n",
    "                if i_gene_type == 0: # First line\n",
    "                    ax.set_title(other_type, fontsize='x-large')\n",
    "                if i_other_type == 0:\n",
    "                    ax.set_ylabel(gene_type, rotation='horizontal', ha='right', fontsize='x-large')\n",
    "\n",
    "    plt.savefig(plot_name, dpi=300, bbox_inches='tight')\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_gene_stats(evol_stats, count_bars=True, plot_name=exp_path.joinpath('gene_pair_counts.png'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_gene_stats(evol_stats, count_bars=False, plot_name=exp_path.joinpath('gene_pair_distances.png'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_gene_stats(neutral_stats, count_bars=True, plot_name=neutral_exp_path.joinpath('gene_pair_counts.png'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plot_gene_stats(neutral_stats, count_bars=False, plot_name=neutral_exp_path.joinpath('gene_pair_distances.png'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot intergene distribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_intergene_distances(exp_path, gen):\n",
    "    rep_dirs = sorted([d for d in exp_path.iterdir() if (d.is_dir() and d.name.startswith(\"rep\"))])\n",
    "    full_res = pd.DataFrame()\n",
    "    for rep, rep_dir in enumerate(rep_dirs):\n",
    "\n",
    "        best_indiv = get_best_indiv(rep_dir, gen)\n",
    "        \n",
    "        intergenes = np.zeros(best_indiv.nb_genes)\n",
    "        for i_gene, gene in enumerate(best_indiv.genes):\n",
    "            intergenes[i_gene] = gene.intergene\n",
    "        indiv_res = pd.DataFrame(data={'Intergene':intergenes}, dtype=int)\n",
    "        indiv_res.insert(0, 'Replicate', rep)\n",
    "\n",
    "        full_res = pd.concat([full_res, indiv_res])\n",
    "    \n",
    "    return full_res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_intergenes(exp_path, neutral_exp_path, gen, cutoff, plot_name):\n",
    "    \n",
    "    intergenes = get_intergene_distances(exp_path, gen)\n",
    "    neutral_intergenes = get_intergene_distances(neutral_exp_path, 100000)\n",
    "    \n",
    "    bins = np.linspace(0.0, 4.0, 40)\n",
    "    \n",
    "    data = [intergenes, neutral_intergenes]\n",
    "    titles = ['With selection', 'Without selection']\n",
    "    \n",
    "    plt.figure(figsize=(9, 4), dpi=300)\n",
    "\n",
    "    for i in range(2):\n",
    "\n",
    "        plt.subplot(1, 2, i+1)\n",
    "        plt.ylim(0, 1)\n",
    "        plt.xlim(-0.2, 4.2)\n",
    "        plt.xlabel('Distance (log)')\n",
    "        plt.ylabel('Density')\n",
    "        # Plot intergene distances\n",
    "        plt.hist(np.log10(data[i]['Intergene']), bins=bins, density=True)\n",
    "        # Plot cutoff line\n",
    "        plt.vlines(np.log10(cutoff), 0, 1, linestyle='--', linewidth=1,\n",
    "                   color='tab:red', label='Cutoff distance')\n",
    "        plt.grid(linestyle=':')\n",
    "        plt.title(titles[i])\n",
    "        plt.legend(loc='upper left')\n",
    "        \n",
    "    plt.savefig(plot_name, dpi=300)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_intergenes(exp_path, neutral_exp_path, gen=gen, cutoff=params['interaction_dist'],\n",
    "                plot_name=exp_path.joinpath('intergene_distr.png'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot local sigma"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sigma_tsc(indiv, sigma_env: float):\n",
    "    temporal_expr = np.zeros((indiv.nb_genes, indiv.nb_eval_steps))\n",
    "\n",
    "    # Initial values at t = 0\n",
    "    temporal_expr[:, 0] = np.array([gene.basal_expression for gene in indiv.genes])\n",
    "    \n",
    "    sigma_tsc = np.zeros((indiv.nb_genes, indiv.nb_eval_steps))\n",
    "\n",
    "    # Iterate the system\n",
    "    for t in range(1, indiv.nb_eval_steps):\n",
    "        sigma_local = indiv.inter_matrix @ temporal_expr[:, t-1]\n",
    "        sigma_tsc[:, t] = sigma_local\n",
    "        sigma_total = indiv.sigma_basal + sigma_local + sigma_env\n",
    "        temporal_expr[:, t] = 1.0 / (1.0 + np.exp((sigma_total - indiv.sigma_opt)/indiv.epsilon))\n",
    "\n",
    "    return sigma_tsc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_sigma_tsc(indiv, sigma_env, all_sigma=False, name=None):\n",
    "    \n",
    "    colors= ['tab:blue', 'tab:red', 'tab:green'] #AB, A, B\n",
    "    \n",
    "    sigma_tsc = get_sigma_tsc(indiv, sigma_env=sigma_env)\n",
    "        \n",
    "    plt.figure(figsize=(9, 6), dpi=300)\n",
    "    for i_gene in range(indiv.nb_genes):\n",
    "        tsc_val = sigma_tsc[i_gene]\n",
    "        if all_sigma:\n",
    "            tsc_val += indiv.sigma_basal + sigma_env\n",
    "        plt.plot(tsc_val, color=colors[indiv.genes[i_gene].gene_type])\n",
    "    \n",
    "    \n",
    "    # Add horizontal activation line\n",
    "    x_min, x_max = plt.xlim()\n",
    "    plt.hlines(params['sigma_opt'], x_min, x_max, linestyle='--', linewidth=1,\n",
    "           color='tab:red', label='Half activation threshold')\n",
    "    plt.xlim(x_min, x_max)\n",
    "\n",
    "    plt.ylim(1, -1)\n",
    "    \n",
    "    plt.xlabel('Time')\n",
    "    if all_sigma:\n",
    "        plt.ylabel('Total supercoiling')\n",
    "    else:\n",
    "        plt.ylabel('TSC supercoiling')\n",
    "    plt.title(f'Environment SC: {sigma_env:.2}')\n",
    "    \n",
    "    plt.grid(linestyle=':')\n",
    "    \n",
    "    if name:\n",
    "        plt.savefig(name, dpi=300, bbox_inches='tight')\n",
    "    \n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_avg_sigma_tsc(indiv, sigma_env, all_sigma=False, title=None, name=None):\n",
    "    \n",
    "    colors = ['tab:blue', 'tab:red', 'tab:green'] #AB, A, B\n",
    "    \n",
    "    sigma_tsc = get_sigma_tsc(indiv, sigma_env=sigma_env)\n",
    "    \n",
    "    avg_sigma_tsc = np.zeros((3, indiv.nb_eval_steps))\n",
    "        \n",
    "    fig, ax = plt.subplots(figsize=(9, 4), dpi=300)\n",
    "    for i_gene in range(indiv.nb_genes):\n",
    "        avg_sigma_tsc[indiv.genes[i_gene].gene_type] += sigma_tsc[i_gene]\n",
    "    \n",
    "    avg_sigma_tsc /= (indiv.nb_genes // 3)\n",
    "    \n",
    "    if all_sigma:\n",
    "        avg_sigma_tsc += indiv.sigma_basal + sigma_env\n",
    "        \n",
    "    for i_gene_type, gene_type in enumerate(gene_types):\n",
    "        ax.plot(avg_sigma_tsc[i_gene_type], color=colors[i_gene_type], label=f'{gene_type} genes')\n",
    "    \n",
    "    ax.set_xlabel('Time')\n",
    "    \n",
    "    # Add horizontal activation line\n",
    "    x_min, x_max = ax.get_xlim()\n",
    "    ax.hlines(params['sigma_opt'], x_min, x_max, linestyle='--', linewidth=1,\n",
    "           color='tab:red', label='Half activation threshold')\n",
    "    ax.set_xlim(x_min, x_max)\n",
    "\n",
    "    \n",
    "    if not all_sigma:\n",
    "        ax.set_ylabel('TSC supercoiling')\n",
    "    else:\n",
    "        ax.set_ylabel('Total supercoiling')\n",
    "        \n",
    "    ax.set_ylim(0.25, -0.85)\n",
    "    \n",
    "    ax.grid(linestyle=':')\n",
    "    ax.legend(loc='upper left')\n",
    "    \n",
    "    if title:\n",
    "        ax.set_title(title)\n",
    "    \n",
    "    if name:\n",
    "        plt.savefig(name, dpi=300, bbox_inches='tight')\n",
    "    \n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_med_sigma_tsc(indiv, sigma_env, all_sigma=False, title=None, name=None):\n",
    "    \n",
    "    colors = ['tab:blue', 'tab:red', 'tab:green'] #AB, A, B\n",
    "    \n",
    "    sigma_tsc = get_sigma_tsc(indiv, sigma_env=sigma_env)\n",
    "    \n",
    "    sigma_tsc_type = {}\n",
    "    \n",
    "    for i_gene_type, gene_type in enumerate(gene_types):\n",
    "        data = [sigma_tsc[i_gene, :] for i_gene in range(indiv.nb_genes)\n",
    "                if indiv.genes[i_gene].gene_type == i_gene_type]\n",
    "        sigma_tsc_type[gene_type] = pd.DataFrame(data)\n",
    "        \n",
    "        if all_sigma:\n",
    "            sigma_tsc_type[gene_type] += indiv.sigma_basal + sigma_env\n",
    "        \n",
    "\n",
    "    fig, ax = plt.subplots(figsize=(9, 4), dpi=300)\n",
    "    \n",
    "    for i_gene_type, gene_type in enumerate(gene_types):\n",
    "        med_sigma_tsc = sigma_tsc_type[gene_type].median()\n",
    "        ax.plot(med_sigma_tsc, color=gene_type_color[i_gene_type], label=f'{gene_type} genes')\n",
    "    \n",
    "    ax.set_xlabel('Time')\n",
    "    \n",
    "    # Add horizontal activation line\n",
    "    x_min, x_max = ax.get_xlim()\n",
    "    ax.hlines(params['sigma_opt'], x_min, x_max, linestyle='--', linewidth=1,\n",
    "           color='tab:red', label='Half activation threshold')\n",
    "    ax.set_xlim(x_min, x_max)\n",
    "\n",
    "    \n",
    "    if not all_sigma:\n",
    "        ax.set_ylabel('TSC supercoiling')\n",
    "    else:\n",
    "        ax.set_ylabel('Total supercoiling')\n",
    "        \n",
    "    ax.set_ylim(0.25, -0.85)\n",
    "    \n",
    "    ax.grid(linestyle=':')\n",
    "    ax.legend(loc='upper left')\n",
    "    \n",
    "    if title:\n",
    "        ax.set_title(title)\n",
    "    \n",
    "    if name:\n",
    "        plt.savefig(name, dpi=300, bbox_inches='tight')\n",
    "    \n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_violin_sigma_tsc(indiv, sigma_env, wanted_type, all_sigma=False, title=None, name=None):\n",
    "        \n",
    "    sigma_tsc = get_sigma_tsc(indiv, sigma_env=sigma_env)\n",
    "    \n",
    "    sigma_tsc_type = {}\n",
    "    \n",
    "    for i_gene_type, gene_type in enumerate(gene_types):\n",
    "        data = [sigma_tsc[i_gene, :] for i_gene in range(indiv.nb_genes)\n",
    "                if indiv.genes[i_gene].gene_type == i_gene_type]\n",
    "        sigma_tsc_type[gene_type] = pd.DataFrame(data)\n",
    "        \n",
    "        if all_sigma:\n",
    "            sigma_tsc_type[gene_type] += indiv.sigma_basal + sigma_env\n",
    "\n",
    "            \n",
    "    fig, ax = plt.subplots(figsize=(9, 4), dpi=300)\n",
    "    \n",
    "    # On plotte que les gènes de type `wanted_type`\n",
    "    for i_gene_type, gene_type in enumerate(gene_types):\n",
    "        if gene_type != wanted_type:\n",
    "            continue\n",
    "    \n",
    "        ax.plot(sigma_tsc_type[gene_type].median(), \n",
    "                #linestyle='',\n",
    "                #marker='x',\n",
    "                color=gene_type_color[i_gene_type], label=f'{gene_type} genes')\n",
    "        \n",
    "    #for i_gene_type, gene_type in enumerate(gene_types):\n",
    "    ax.violinplot(sigma_tsc_type[wanted_type], positions=np.arange(indiv.nb_eval_steps),\n",
    "                  showmedians=True) # A genes only for now\n",
    "    \n",
    "    ax.set_xlabel('Time')\n",
    "    \n",
    "    # Add horizontal activation line\n",
    "    x_min, x_max = ax.get_xlim()\n",
    "    ax.hlines(params['sigma_opt'], x_min, x_max, linestyle='--', linewidth=1,\n",
    "           color='tab:red', label='Half activation threshold')\n",
    "    ax.set_xlim(x_min, x_max)\n",
    "\n",
    "    \n",
    "    if not all_sigma:\n",
    "        ax.set_ylabel('TSC supercoiling')\n",
    "    else:\n",
    "        ax.set_ylabel('Total supercoiling')\n",
    "        \n",
    "    ax.set_ylim(0.85, -0.85)\n",
    "    \n",
    "    if title:\n",
    "        ax.set_title(title)\n",
    "    ax.grid(linestyle=':')\n",
    "    ax.legend(loc='upper left')\n",
    "    \n",
    "    if name:\n",
    "        plt.savefig(name, dpi=300, bbox_inches='tight')\n",
    "    \n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_best_sigma_tsc(exp_path, gen):\n",
    "    rep_dirs = sorted([d for d in exp_path.iterdir() if (d.is_dir() and d.name.startswith(\"rep\"))])\n",
    "    \n",
    "    for rep, rep_dir in enumerate(rep_dirs):\n",
    "\n",
    "        best_indiv = get_best_indiv(rep_dir, gen)\n",
    "                \n",
    "        plot_violin_sigma_tsc(best_indiv, sigma_env=params['sigma_A'], wanted_type='A', all_sigma=True,\n",
    "                              title='Environment A: A genes',\n",
    "                              name=exp_path.joinpath(f'violin_sc_rep{rep:02}_env_A_genes_A.png'))\n",
    "        plot_violin_sigma_tsc(best_indiv, sigma_env=params['sigma_A'], wanted_type='B', all_sigma=True,\n",
    "                              title='Environment A: B genes',\n",
    "                              name=exp_path.joinpath(f'violin_sc_rep{rep:02}_env_A_genes_B.png'))\n",
    "        plot_violin_sigma_tsc(best_indiv, sigma_env=params['sigma_A'], wanted_type='AB', all_sigma=True,\n",
    "                              title='Environment A: AB genes',\n",
    "                              name=exp_path.joinpath(f'violin_sc_rep{rep:02}_env_A_genes_AB.png'))\n",
    "\n",
    "        \n",
    "        #plot_sigma_tsc(best_indiv, sigma_env=params['sigma_A'], all_sigma=True,\n",
    "        #               name=exp_path.joinpath(f'sigma_tsc_best_rep{rep:02}_env_A.png'))\n",
    "    \n",
    "        #plot_sigma_tsc(best_indiv, sigma_env=params['sigma_B'], all_sigma=True,\n",
    "        #               name=exp_path.joinpath(f'sigma_tsc_best_rep{rep:02}_env_B.png'))\n",
    "\n",
    "        #plot_avg_sigma_tsc(best_indiv, sigma_env=params['sigma_A'], title='Environment A: mean SC',\n",
    "        #                   name=exp_path.joinpath(f'sigma_tsc_avg_best_rep{rep:02}_env_A_mean.png'))\n",
    "        #plot_med_sigma_tsc(best_indiv, sigma_env=params['sigma_A'], title='Environment A: median SC',\n",
    "        #                   name=exp_path.joinpath(f'sigma_tsc_avg_best_rep{rep:02}_env_A_med.png'))\n",
    "\n",
    "        #plot_avg_sigma_tsc(best_indiv, sigma_env=params['sigma_B'],\n",
    "        #                   name=exp_path.joinpath(f'sigma_tsc_avg_best_rep{rep:02}_env_B.png'))\n",
    "        \n",
    "        #plot_avg_sigma_tsc(best_indiv, sigma_env=params['sigma_A'], all_sigma=True,\n",
    "        #                   title='Environment A: mean SC',\n",
    "        #                   name=exp_path.joinpath(f'sigma_tsc_all_best_rep{rep:02}_env_A_mean.png'))\n",
    "        plot_med_sigma_tsc(best_indiv, sigma_env=params['sigma_A'], all_sigma=True,\n",
    "                           title='Environment A: median SC',\n",
    "                           name=exp_path.joinpath(f'sigma_tsc_all_best_rep{rep:02}_env_A_med.png'))\n",
    "\n",
    "        #plot_avg_sigma_tsc(best_indiv, sigma_env=params['sigma_B'], all_sigma=True, title='Environment B',\n",
    "        #                   name=exp_path.joinpath(f'sigma_tsc_all_best_rep{rep:02}_env_B.png'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_best_sigma_tsc(exp_path, gen)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Detailed plot of every gene of a type in an individual (for debugging)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_every_gene(indiv, sigma_env, wanted_type, base_name, extra):\n",
    "    \n",
    "    sigma_tsc = get_sigma_tsc(indiv, sigma_env=sigma_env)\n",
    "    sigma_total = sigma_tsc + indiv.sigma_basal + sigma_env\n",
    "    temporal_expr = indiv.run_system(sigma_env)\n",
    "        \n",
    "    i_gene_type = 0 # we skip genes of the wrong type but want to count uniformly\n",
    "    for i_gene, gene in enumerate(indiv.genes):\n",
    "    \n",
    "        if gene_types[gene.gene_type] == wanted_type:\n",
    "            \n",
    "            print(f'Gene {i_gene} type {gene_types[gene.gene_type]}: '\n",
    "                  f'mean activation {np.sum(temporal_expr[i_gene, :])/indiv.nb_eval_steps:.2}') \n",
    "\n",
    "            \n",
    "\n",
    "            fig, ax1 = plt.subplots(figsize=(9, 4), dpi=300)\n",
    "\n",
    "            ## Axis 1: Total supercoiling\n",
    "            ax1.set_ylim(0.65, -0.85)\n",
    "            ax1.set_ylabel('Total supercoiling', color='tab:red')\n",
    "            ax1.set_xlabel('Time')\n",
    "            ax1.grid(linestyle=':')\n",
    "\n",
    "            ax1.plot(sigma_total[i_gene, :], color='tab:red', label='SC level')\n",
    "            \n",
    "            x_min, x_max = ax1.get_xlim()\n",
    "            ax1.hlines(params['sigma_opt'], x_min, x_max, linestyle='--', linewidth=1,\n",
    "                   color='tab:red', label='Half activation (SC)')\n",
    "            ax1.set_xlim(x_min, x_max)\n",
    "\n",
    "\n",
    "            ## Axis 2: Expression level\n",
    "            ax2 = ax1.twinx()\n",
    "            ax2.set_ylim(-0.02, 1.02)\n",
    "            ax2.set_ylabel('Expression level', color='tab:green')\n",
    "\n",
    "            ax2.plot(temporal_expr[i_gene, :], linestyle='--',\n",
    "                     color='tab:green', label='Expression level')\n",
    "            \n",
    "            x_min, x_max = ax2.get_xlim()\n",
    "            ax2.hlines(0.5, x_min, x_max, linestyle='--', linewidth=1,\n",
    "                   color='tab:green', label='Half activation (expr)')\n",
    "            ax2.set_xlim(x_min, x_max)\n",
    "\n",
    "            nb_steps = 5\n",
    "            mean_activ = np.sum(temporal_expr[i_gene, indiv.nb_eval_steps-nb_steps:]) / nb_steps\n",
    "            if mean_activ > 0.5:\n",
    "                plt.title(f'Gene {i_gene} type {wanted_type}: activated')\n",
    "            else:\n",
    "                plt.title(f'Gene {i_gene} type {wanted_type}: inhibited')\n",
    "\n",
    "            fig.legend(loc=\"upper left\", bbox_to_anchor=(0,1), bbox_transform=ax1.transAxes)\n",
    "            \n",
    "            plt.savefig(base_name.joinpath(f'plot_all_genes_{wanted_type}_gene_{i_gene}_{extra}.png'), dpi=300)\n",
    "            \n",
    "            plt.close()\n",
    "            i_gene_type += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plot_every_gene(focal_indiv, sigma_env=params['sigma_A'], wanted_type='B',\n",
    "#                base_name=exp_path, extra='before')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot average supercoiling for each type of gene over replicas (_il faut changer les moyennes en médianes_)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_avg_sigma(exp_path, gen):\n",
    "    \n",
    "    rep_dirs = sorted([d for d in exp_path.iterdir() if (d.is_dir() and d.name.startswith(\"rep\"))])\n",
    "    nb_reps = len(rep_dirs)\n",
    "    \n",
    "    time_steps = 5\n",
    "    \n",
    "    full_data = pd.DataFrame()\n",
    "\n",
    "    for rep, rep_dir in enumerate(rep_dirs):\n",
    "        best_indiv = get_best_indiv(rep_dir, gen)\n",
    "        \n",
    "        sigma_A = get_sigma_tsc(best_indiv, sigma_env=params['sigma_A'])\n",
    "        sigma_B = get_sigma_tsc(best_indiv, sigma_env=params['sigma_B'])\n",
    "\n",
    "        # Take temporal averages\n",
    "        mean_sigma_A_gene = np.sum(sigma_A[:, best_indiv.nb_eval_steps-time_steps:], axis=1) / time_steps\n",
    "        mean_sigma_B_gene = np.sum(sigma_B[:, best_indiv.nb_eval_steps-time_steps:], axis=1) / time_steps\n",
    "        \n",
    "        # Sort by gene type\n",
    "        mean_sigma_A = np.zeros(3)\n",
    "        mean_sigma_B = np.zeros(3)\n",
    "        for i_gene in range(best_indiv.nb_genes):\n",
    "            mean_sigma_A[best_indiv.genes[i_gene].gene_type] += mean_sigma_A_gene[i_gene]\n",
    "            mean_sigma_B[best_indiv.genes[i_gene].gene_type] += mean_sigma_B_gene[i_gene]\n",
    "\n",
    "        # Divide by number of genes in each class\n",
    "        mean_sigma_A /= (best_indiv.nb_genes // 3)\n",
    "        mean_sigma_B /= (best_indiv.nb_genes // 3)\n",
    "        \n",
    "        # Add other supercoiling sources\n",
    "        mean_sigma_A += best_indiv.sigma_basal + params['sigma_A']\n",
    "        mean_sigma_B += best_indiv.sigma_basal + params['sigma_B']\n",
    "        \n",
    "        col_names = []\n",
    "        \n",
    "        for gene_type in gene_types:\n",
    "            col_names.append(f'{gene_type}sc_A')\n",
    "            \n",
    "        for gene_type in gene_types:\n",
    "            col_names.append(f'{gene_type}sc_B')\n",
    "                    \n",
    "        rep_data = pd.DataFrame(data=[np.concatenate([mean_sigma_A, mean_sigma_B])], \n",
    "                                columns=col_names)\n",
    "        \n",
    "        rep_data.insert(0, 'Replicate', rep)\n",
    "        \n",
    "        full_data = pd.concat([full_data, rep_data])\n",
    "\n",
    "    full_data = full_data.set_index('Replicate')\n",
    "\n",
    "    return full_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_mean_sigma(exp_path, gen, title=None):\n",
    "    \n",
    "    full_data = get_avg_sigma(exp_path, gen)\n",
    "                \n",
    "    ## Actual plotting\n",
    "    mean_stats = full_data.mean()\n",
    "    std_stats = full_data.std()\n",
    "    med_stats = full_data.median()\n",
    "    \n",
    "    fig, ax = plt.subplots(figsize=(9, 4), dpi=300)\n",
    "\n",
    "    x = np.arange(3)  # 3 types of genes\n",
    "    width = 0.1  # the width of the bars\n",
    "\n",
    "    env_A_means = np.array([mean_stats[\"ABsc_A\"], mean_stats[\"Asc_A\"], mean_stats[\"Bsc_A\"]])\n",
    "    env_B_means = np.array([mean_stats[\"ABsc_B\"], mean_stats[\"Asc_B\"], mean_stats[\"Bsc_B\"]])\n",
    "\n",
    "    env_A_std = np.array([std_stats[\"ABsc_A\"], std_stats[\"Asc_A\"], std_stats[\"Bsc_A\"]])\n",
    "    env_B_std = np.array([std_stats[\"ABsc_B\"], std_stats[\"Asc_B\"], std_stats[\"Bsc_B\"]])\n",
    "    \n",
    "    env_A_med = np.array([med_stats[\"ABsc_A\"], med_stats[\"Asc_A\"], med_stats[\"Bsc_A\"]])\n",
    "    env_B_med = np.array([med_stats[\"ABsc_B\"], med_stats[\"Asc_B\"], med_stats[\"Bsc_B\"]])\n",
    "\n",
    "\n",
    "    # Plot mean and std\n",
    "    ax.errorbar(x - width/2,\n",
    "                env_A_means,\n",
    "                label='Environment A',\n",
    "                marker='o',\n",
    "                linestyle='',\n",
    "                color='tab:blue',\n",
    "                yerr=env_A_std,\n",
    "                capsize=5)\n",
    "\n",
    "    ax.errorbar(x + width/2,\n",
    "                env_B_means,\n",
    "                marker='o',\n",
    "                linestyle='',\n",
    "                color='tab:orange',\n",
    "                label='Environment B',\n",
    "                yerr=env_B_std,\n",
    "                capsize=5)\n",
    "    \n",
    "    # Add median values -------- note : c'est la médiane des *moyennes*, or ce sont les *médianes* qu'il faut\n",
    "    ax.plot(x - width/2,\n",
    "            env_A_med,\n",
    "            marker='x',\n",
    "            linestyle='',\n",
    "            color='tab:blue')\n",
    "    \n",
    "    ax.plot(x + width/2,\n",
    "            env_B_med,\n",
    "            marker='x',\n",
    "            linestyle='',\n",
    "            color='tab:orange')\n",
    "    \n",
    "    \n",
    "    # Add half-activation threshold (sigma_opt)\n",
    "    x_min, x_max = ax.get_xlim()\n",
    "    ax.hlines(params['sigma_opt'], x_min, x_max, linestyle='--', linewidth=1,\n",
    "           color='tab:red', label='Half activation threshold')\n",
    "    ax.set_xlim(x_min, x_max)\n",
    "\n",
    "    ax.set_ylabel('SC level', fontsize=label_fontsize)\n",
    "    ax.set_ylim(0.25, -0.85)\n",
    "    ax.set_xticks(x)\n",
    "    ax.set_xticklabels([\"AB genes\", \"A genes\", \"B genes\"])\n",
    "\n",
    "    plt.grid(linestyle=':', axis='y')\n",
    "\n",
    "    plt.tick_params(axis='both', which='major', labelsize=tick_fontsize)\n",
    "\n",
    "    plt.legend(fontsize=legend_fontsize, loc='upper center')\n",
    "    \n",
    "    if title:\n",
    "        plt.title(title, fontsize=label_fontsize)\n",
    "\n",
    "    plt.savefig(f'{exp_path}/mean_supercoiling.png', bbox_inches='tight')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_mean_sigma(exp_path, gen, title=\"Evolve basal SC\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_mean_sigma(pathlib.Path('/Users/theotime/Desktop/evotsc/new_100k/'), 100_000, title=\"Vanilla\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_mean_sigma(pathlib.Path('/Users/theotime/Desktop/evotsc/unbiased_gen_size_750k/'), 750_000, title=\"Evolve genome size\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot gene expression levels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_avg_expr(indiv, sigma_env, plot_title, plot_name):\n",
    "\n",
    "    temporal_expr = indiv.run_system(sigma_env)\n",
    "\n",
    "    nb_genes, nb_steps = temporal_expr.shape\n",
    "\n",
    "    colors = mpl.cm.get_cmap('viridis', nb_genes)(range(nb_genes))\n",
    "\n",
    "    plt.figure(figsize=(9, 4), dpi=dpi)\n",
    "\n",
    "    plt.ylim(-0.05, 1.05)\n",
    "    \n",
    "    type_expr = np.zeros((3, nb_steps))\n",
    "    \n",
    "    for wanted_gene_type in gene_types:\n",
    "        for i_gene, gene in enumerate(indiv.genes):\n",
    "            if gene_types[gene.gene_type] == wanted_gene_type:\n",
    "                type_expr[gene.gene_type] += temporal_expr[i_gene, :]\n",
    "        \n",
    "    type_expr /= (nb_genes // 3)\n",
    "\n",
    "    for i_gene_type, gene_type in enumerate(gene_types):\n",
    "        plt.plot(type_expr[i_gene_type, :],\n",
    "                 color=gene_type_color[i_gene_type],\n",
    "                 label=f'{gene_type} genes')\n",
    "\n",
    "    plt.grid(linestyle=':')\n",
    "    plt.xlabel('Time', fontsize='large')\n",
    "    plt.ylabel('Expression level', fontsize='large')\n",
    "    \n",
    "    x_min, x_max = plt.xlim()\n",
    "    plt.hlines(0.5, x_min, x_max, linestyle='--', linewidth=1,\n",
    "           color='tab:red', label='Half activation threshold')\n",
    "    plt.xlim(x_min, x_max)\n",
    "\n",
    "\n",
    "    plt.legend(loc='lower left')\n",
    "    plt.title(plot_title)\n",
    "    \n",
    "    plt.savefig(exp_path.joinpath(plot_name), dpi=300, bbox_inches='tight')\n",
    "\n",
    "    plt.show()\n",
    "\n",
    "    plt.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_avg_expr_all(exp_path, gen):\n",
    "    rep_dirs = sorted([d for d in exp_path.iterdir() if (d.is_dir() and d.name.startswith(\"rep\"))])\n",
    "    \n",
    "    for rep, rep_dir in enumerate(rep_dirs):\n",
    "        \n",
    "        indiv = get_best_indiv(rep_dir, gen)\n",
    "        \n",
    "        plot_avg_expr(indiv, params['sigma_A'], 'Environment A', f'avg_expr_rep{rep:02}_env_A.png')\n",
    "        #plot_avg_expr(indiv, params['sigma_B'], 'Environment B', f'avg_expr_rep{rep:02}_env_B.png')\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_avg_expr_all(exp_path, gen)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
